beacon_node/beacon_chain/src/validator_custody.rs

```diff
@@ -0,0 +1,433 @@
+        ssz_context: CustodyContextSsz,
+        is_supernode: bool,
+    ) -> Self {
+        let (sender, _) = channel(CHANNEL_CAPACITY);
+        CustodyContext {
+            advertised_validator_custody_count: AtomicU64::new(
+                ssz_context.advertised_validator_custody,
+            ),
+            validator_custody_count_at_head: AtomicU64::new(ssz_context.validator_custody_at_head),
+            current_is_supernode: is_supernode,
+            persisted_is_supernode: ssz_context.persisted_is_supernode,
+            validator_registrations: Default::default(),
+            sender,
+        }
+    }
+
+    pub fn subscribe(&self) -> Receiver<CustodyContextMessage> {
+        self.sender.subscribe()
+    }
+
+    /// Register a new validator index and updates the list of validators if required.
+    ///
+    /// Also modifies the internal structures if the validator custody has changed to
+    /// update the `custody_column_count`.
+    pub fn register_validator<E: EthSpec>(
+        &self,
+        validators_and_balance: Vec<(usize, u64)>,
+        slot: Slot,
+        spec: &ChainSpec,
+    ) {
+        // Complete all registrations
+        let mut registrations = self.validator_registrations.write();
+        for (validator_index, effective_balance) in validators_and_balance {
+            registrations.register_validator::<E>(validator_index, effective_balance, slot, spec);
+        }
+
+        // Completed registrations, now check if the validator custody requirement has changed
+        let Some(new_validator_custody) = registrations.latest_validator_custody_requirement()
+        else {
+            return;
+        };
+
+        let current_cgc = self.custody_group_count(spec);
+        let validator_custody_count_at_head =
+            self.validator_custody_count_at_head.load(Ordering::Relaxed);
+
+        if new_validator_custody != validator_custody_count_at_head {
+            tracing::debug!(
+                old_count = validator_custody_count_at_head,
+                new_count = new_validator_custody,
+                "Validator count at head updated"
+            );
+            self.validator_custody_count_at_head
+                .store(new_validator_custody, Ordering::Relaxed);
+
+            let updated_cgc = self.custody_group_count(spec);
+            // Send the message to network only if there are more columns subnets to subscribe to
+            if updated_cgc > current_cgc {
+                tracing::debug!(
+                    old_cgc = current_cgc,
+                    updated_cgc,
+                    "Custody group count updated"
+                );
+                if let Err(e) = self
+                    .sender
+                    .send(CustodyContextMessage::HeadCustodyCountChanged {
+                        new_custody_count: updated_cgc,
+                    })
+                {
+                    tracing::error!(error=?e, "Failed to send custody context message");
```

Comment:
How do we handle this correctly?
What if the `send` fails, and the message would never be sent even though the context is updated, and it won't be retried again until the `new_validator_custody` changes again.
Should we flip the order and only update after successfully send?

Reply:
The send would only fail if there are no receivers. I think since we atleast have one receiver in the network service, this should never fail because the network service loop should ideally start before the http service and if it doesn't, I think its good if we detect it early so we can change some assumptions here regarding ordering of operations.